paywall/src/components/Paywall.js

```diff
@@ -7,69 +7,48 @@ import DeveloperOverlay from './developer/DeveloperOverlay'
 import ShowWhenLocked from './lock/ShowWhenLocked'
 import ShowWhenUnlocked from './lock/ShowWhenUnlocked'
 import GlobalErrorProvider from '../utils/GlobalErrorProvider'
-import { lockPage, unlockPage } from '../services/iframeService'
 import { UnlockedFlag } from './lock/UnlockFlag'
 import { lockRoute } from '../utils/routes'
+import useListenForPostMessage from '../hooks/browser/useListenForPostMessage'
+import usePostMessage from '../hooks/browser/usePostMessage'
 
-class Paywall extends React.Component {
-  constructor(props) {
-    super(props)
-    this.state = {
-      scrollPosition: 0,
-    }
-    this.handleScrollPosition = this.handleScrollPosition.bind(this)
-    this.handleIframe = this.handleIframe.bind(this)
-  }
-  componentDidMount() {
-    window.addEventListener('message', event => {
-      if (event.data.scrollPosition) {
-        this.handleScrollPosition(event.data.scrollPosition)
+export function Paywall({ locks, locked, redirect, window }) {
+  const data = useListenForPostMessage(window) || { scrollPosition: 0 }
+  const scrollPosition = data.scrollPosition || 0
+  const { postMessage } = usePostMessage(window)
+  useEffect(
+    () => {
+      if (locked) {
+        postMessage('locked')
+      } else {
+        postMessage('unlocked')
       }
-    })
-    this.handleIframe()
-  }
-  componentDidUpdate() {
-    this.handleIframe()
-  }
-  handleScrollPosition(scrollPosition) {
-    this.setState(state => {
-      if (state.scrollPosition === scrollPosition) return null
-      return { scrollPosition }
-    })
-  }
-  handleIframe() {
-    const { locked } = this.props
-    if (locked) {
-      lockPage()
-    } else {
-      unlockPage()
-    }
-  }
-  render() {
-    const { scrollPosition } = this.state
-    const { locks, locked, redirect } = this.props
-    return (
-      <GlobalErrorProvider>
-        <ShowWhenLocked locked={locked}>
-          <Overlay
-            scrollPosition={scrollPosition}
-            locks={locks}
-            redirect={redirect}
-          />
-          <DeveloperOverlay />
-        </ShowWhenLocked>
-        <ShowWhenUnlocked locked={locked}>
-          <UnlockedFlag />
-        </ShowWhenUnlocked>
-      </GlobalErrorProvider>
-    )
-  }
+    },
+    [locked]
+  )
+
+  return (
+    <GlobalErrorProvider>
+      <ShowWhenLocked locked={locked}>
+        <Overlay
+          scrollPosition={scrollPosition}
+          locks={locks}
+          redirect={redirect}
+        />
+        <DeveloperOverlay />
+      </ShowWhenLocked>
+      <ShowWhenUnlocked locked={locked}>
+        <UnlockedFlag />
+      </ShowWhenUnlocked>
+    </GlobalErrorProvider>
+  )
 }
 
 Paywall.propTypes = {
   locks: PropTypes.arrayOf(UnlockPropTypes.lock).isRequired,
   locked: PropTypes.bool.isRequired,
   redirect: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
+  window: PropTypes.shape({ postMessage: PropTypes.func }).isRequired, // for ease of unit testing
```

Comment:
:(

Reply:
This still feels very strange.

Reply:
I'd love it if you could some research on actual good practices for this kind of dependency injection.